// @coreui/react/dist/esm/components/grid/CCol.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var BREAKPOINTS = [
    'xxl',
    'xl',
    'lg',
    'md',
    'sm',
    'xs',
];
var CCol = forwardRef(function (_a, ref) {
    var children = _a.children, className = _a.className, rest = __rest(_a, ["children", "className"]);
    var repsonsiveClassNames = [];
    BREAKPOINTS.forEach(function (bp) {
        var breakpoint = rest[bp];
        delete rest[bp];
        var infix = bp === 'xs' ? '' : "-".concat(bp);
        if (typeof breakpoint === 'number' || typeof breakpoint === 'string') {
            repsonsiveClassNames.push("col".concat(infix, "-").concat(breakpoint));
        }
        if (typeof breakpoint === 'boolean') {
            repsonsiveClassNames.push("col".concat(infix));
        }
        if (breakpoint && typeof breakpoint === 'object') {
            if (typeof breakpoint.span === 'number' || typeof breakpoint.span === 'string') {
                repsonsiveClassNames.push("col".concat(infix, "-").concat(breakpoint.span));
            }
            if (typeof breakpoint.span === 'boolean') {
                repsonsiveClassNames.push("col".concat(infix));
            }
            if (typeof breakpoint.order === 'number' || typeof breakpoint.order === 'string') {
                repsonsiveClassNames.push("order".concat(infix, "-").concat(breakpoint.order));
            }
            if (typeof breakpoint.offset === 'number') {
                repsonsiveClassNames.push("offset".concat(infix, "-").concat(breakpoint.offset));
            }
        }
    });
    var _className = classNames(repsonsiveClassNames.length ? repsonsiveClassNames : 'col', className);
    return (React__default.createElement("div", __assign({ className: _className }, rest, { ref: ref }), children));
});
var span = propTypesExports.oneOfType([
    propTypesExports.bool,
    propTypesExports.number,
    propTypesExports.string,
    propTypesExports.oneOf(['auto']),
]);
var col = propTypesExports.oneOfType([
    span,
    propTypesExports.shape({
        span: span,
        offset: propTypesExports.oneOfType([propTypesExports.number, propTypesExports.string]),
        order: propTypesExports.oneOfType([
            propTypesExports.oneOf(['first', 'last']),
            propTypesExports.number,
            propTypesExports.string,
        ]),
    }),
]);
CCol.propTypes = {
    children: propTypesExports.node,
    className: propTypesExports.string,
    xs: col,
    sm: col,
};
CCol.displayName = 'CCol';

export { CCol };
