// @coreui/react/dist/esm/components/image/CImage.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CImage = forwardRef(function (_a, ref) {
    var _b;
    var align = _a.align, className = _a.className, fluid = _a.fluid, rounded = _a.rounded, thumbnail = _a.thumbnail, rest = __rest(_a, ["align", "className", "fluid", "rounded", "thumbnail"]);
    var _className = classNames((_b = {},
        _b["float-".concat(align)] = align && (align === 'start' || align === 'end'),
        _b['d-block mx-auto'] = align && align === 'center',
        _b['img-fluid'] = fluid,
        _b.rounded = rounded,
        _b['img-thumbnail'] = thumbnail,
        _b), className);
    return React__default.createElement("img", __assign({ className: _className }, rest, { ref: ref }));
});
CImage.propTypes = {
    align: propTypesExports.oneOf(['start', 'center', 'end']),
    className: propTypesExports.string,
    fluid: propTypesExports.bool,
    rounded: propTypesExports.bool,
    thumbnail: propTypesExports.bool,
};
CImage.displayName = 'CImage';

export { CImage };
