// @coreui/react/dist/esm/components/progress/CProgressBar.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { colorPropType } from '../Types.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CProgressBar = forwardRef(function (_a, ref) {
    var _b;
    var children = _a.children, animated = _a.animated, className = _a.className, color = _a.color, _c = _a.value, value = _c === void 0 ? 0 : _c, variant = _a.variant, rest = __rest(_a, ["children", "animated", "className", "color", "value", "variant"]);
    var _className = classNames('progress-bar', (_b = {},
        _b["bg-".concat(color)] = color,
        _b["progress-bar-".concat(variant)] = variant,
        _b['progress-bar-animated'] = animated,
        _b), className);
    return (React__default.createElement("div", __assign({ className: _className, role: "progressbar", style: { width: "".concat(value, "%") }, "aria-valuenow": value, "aria-valuemin": 0, "aria-valuemax": 100 }, rest, { ref: ref }), children));
});
CProgressBar.propTypes = {
    animated: propTypesExports.bool,
    children: propTypesExports.node,
    className: propTypesExports.string,
    color: colorPropType,
    value: propTypesExports.number,
    variant: propTypesExports.oneOf(['striped']),
};
CProgressBar.displayName = 'CProgressBar';

export { CProgressBar };
