// @coreui/react/dist/esm/components/table/CTableRow.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { colorPropType } from '../Types.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CTableRow = forwardRef(function (_a, ref) {
    var _b;
    var children = _a.children, active = _a.active, align = _a.align, className = _a.className, color = _a.color, rest = __rest(_a, ["children", "active", "align", "className", "color"]);
    var _className = classNames((_b = {},
        _b["align-".concat(align)] = align,
        _b['table-active'] = active,
        _b["table-".concat(color)] = color,
        _b), className);
    return (React__default.createElement("tr", __assign({ className: _className ? _className : undefined }, rest, { ref: ref }), children));
});
CTableRow.propTypes = {
    active: propTypesExports.bool,
    align: propTypesExports.oneOf(['bottom', 'middle', 'top']),
    children: propTypesExports.node,
    className: propTypesExports.string,
    color: colorPropType,
};
CTableRow.displayName = 'CTableRow';

export { CTableRow };
