// @coreui/react/dist/esm/components/widgets/CWidgetStatsB.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import { colorPropType } from '../Types.js';
import { CCard } from '../card/CCard.js';
import { CCardBody } from '../card/CCardBody.js';
import { CProgress } from '../progress/CProgress.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CWidgetStatsB = forwardRef(function (_a, ref) {
    var className = _a.className, color = _a.color, inverse = _a.inverse, progress = _a.progress, text = _a.text, title = _a.title, value = _a.value, rest = __rest(_a, ["className", "color", "inverse", "progress", "text", "title", "value"]);
    return (React__default.createElement(CCard, __assign({ className: className, color: color }, (inverse && { textColor: 'high-emphasis-inverse' }), rest, { ref: ref }),
        React__default.createElement(CCardBody, null,
            value && React__default.createElement("div", { className: "fs-4 fw-semibold" }, value),
            title && React__default.createElement("div", null, title),
            React__default.createElement(CProgress, __assign({ className: "my-2", height: 4 }, (inverse && { white: true }), progress)),
            text && (React__default.createElement("small", { className: inverse ? 'text-medium-emphasis-inverse' : 'text-medium-emphasis' }, text)))));
});
CWidgetStatsB.propTypes = {
    className: propTypesExports.string,
    color: colorPropType,
    inverse: propTypesExports.bool,
    progress: propTypesExports.object,
    text: propTypesExports.string,
    title: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node]),
    value: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node, propTypesExports.number]),
};
CWidgetStatsB.displayName = 'CWidgetCWidgetStatsB';

export { CWidgetStatsB };
