// @coreui/react/dist/esm/components/widgets/CWidgetStatsA.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { colorPropType } from '../Types.js';
import { CCard } from '../card/CCard.js';
import { CCardBody } from '../card/CCardBody.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CWidgetStatsA = forwardRef(function (_a, ref) {
    var _b;
    var action = _a.action, chart = _a.chart, className = _a.className, color = _a.color, title = _a.title, value = _a.value, rest = __rest(_a, ["action", "chart", "className", "color", "title", "value"]);
    var _className = classNames((_b = {}, _b["bg-".concat(color)] = color, _b['text-high-emphasis-inverse'] = color, _b), className);
    return (React__default.createElement(CCard, __assign({ className: _className }, rest, { ref: ref }),
        React__default.createElement(CCardBody, { className: "pb-0 d-flex justify-content-between align-items-start" },
            React__default.createElement("div", null,
                value && React__default.createElement("div", { className: "fs-4 fw-semibold" }, value),
                title && React__default.createElement("div", null, title)),
            action),
        chart));
});
CWidgetStatsA.propTypes = {
    action: propTypesExports.node,
    chart: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node]),
    className: propTypesExports.string,
    color: colorPropType,
    title: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node]),
    value: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node, propTypesExports.number]),
};
CWidgetStatsA.displayName = 'CWidgetStatsA';

export { CWidgetStatsA };
