// @coreui/react/dist/esm/components/widgets/CWidgetStatsD.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { colorPropType } from '../Types.js';
import { CCard } from '../card/CCard.js';
import { CCardBody } from '../card/CCardBody.js';
import { CCardHeader } from '../card/CCardHeader.js';
import { CCol } from '../grid/CCol.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CWidgetStatsD = forwardRef(function (_a, ref) {
    var _b;
    var className = _a.className, chart = _a.chart, color = _a.color, icon = _a.icon, values = _a.values, rest = __rest(_a, ["className", "chart", "color", "icon", "values"]);
    var _className = classNames(className);
    var classNameHeader = classNames('position-relative d-flex justify-content-center align-items-center', (_b = {},
        _b["bg-".concat(color)] = color,
        _b));
    return (React__default.createElement(CCard, __assign({ className: _className }, rest, { ref: ref }),
        React__default.createElement(CCardHeader, { className: classNameHeader },
            icon,
            chart),
        React__default.createElement(CCardBody, { className: "row text-center" }, values &&
            values.map(function (value, index) {
                return (React__default.createElement(React__default.Fragment, { key: index },
                    index % 2 !== 0 && React__default.createElement("div", { className: "vr" }),
                    React__default.createElement(CCol, null,
                        React__default.createElement("div", { className: "fs-5 fw-semibold" }, value.value),
                        React__default.createElement("div", { className: "text-uppercase text-medium-emphasis small" }, value.title))));
            }))));
});
CWidgetStatsD.propTypes = {
    chart: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node]),
    className: propTypesExports.string,
    color: colorPropType,
    icon: propTypesExports.oneOfType([propTypesExports.string, propTypesExports.node]),
    values: propTypesExports.arrayOf(propTypesExports.any),
};
CWidgetStatsD.displayName = 'CWidgetStatsD';

export { CWidgetStatsD };
