// @coreui/react/dist/esm/components/avatar/CAvatar.js
import { __rest, __assign } from '../../node_modules/tslib/tslib.es6.js';
import React__default, { forwardRef } from 'react';
import '../../node_modules/prop-types/index.js';
import classNames from '../../node_modules/classnames/index.js';
import { colorPropType, shapePropType, textColorsPropType } from '../Types.js';
import { exports as propTypesExports } from '../../_virtual/index.js';

var CAvatar = forwardRef(function (_a, ref) {
    var _b;
    var children = _a.children, className = _a.className, color = _a.color, shape = _a.shape, size = _a.size, src = _a.src, status = _a.status, textColor = _a.textColor, rest = __rest(_a, ["children", "className", "color", "shape", "size", "src", "status", "textColor"]);
    var _className = classNames('avatar', (_b = {},
        _b["bg-".concat(color)] = color,
        _b["avatar-".concat(size)] = size,
        _b["text-".concat(textColor)] = textColor,
        _b), shape, className);
    var statusClassName = status && classNames('avatar-status', "bg-".concat(status));
    return (React__default.createElement("div", __assign({ className: _className }, rest, { ref: ref }),
        src ? React__default.createElement("img", { src: src, className: "avatar-img" }) : children,
        status && React__default.createElement("span", { className: statusClassName })));
});
CAvatar.propTypes = {
    children: propTypesExports.node,
    className: propTypesExports.string,
    color: colorPropType,
    shape: shapePropType,
    size: propTypesExports.string,
    src: propTypesExports.string,
    status: propTypesExports.string,
    textColor: textColorsPropType,
};
CAvatar.displayName = 'CAvatar';

export { CAvatar };
